data.md

```diff
@@ -200,5 +216,58 @@ Operator `_++_` implements an append operator for sort `Stack`.
 
     rule #drop(.ValTypes,   STACK)                       => STACK
     rule #drop(TYPE VTYPES, < TYPE > VAL:Number : STACK) => #drop(VTYPES, STACK)
+```
+
+Byte Map
+--------
+
+Wasm memories are byte arrays, sized in pages of 65536 bytes, initialized to be all zero bytes.
+To avoid storing many zeros in what may be sparse memory, we implement memory as maps, and store only non-zero bytes.
+The maps store integers, but maintains the invariant that each stored integer is between 1 and 255, inclusive.
+`BM [ N := I ]` writes the integer `I` to memory as bytes (little-endian), starting at index `N`.
+
+```k
+    syntax Map ::= Map "[" Int ":=" Int "]" [function]
+ // --------------------------------------------------
+    rule BM [ IDX := 0  ] => BM
+    rule BM [ IDX := VAL] =>
+        // Don't store 0 bytes.
+        #if VAL modInt 256 =/=Int 0
+          #then BM [IDX <- VAL modInt 256 ]
+          #else BM [IDX <- undef          ]
+        #fi [ IDX +Int 1 := VAL /Int 256 ]
+      requires VAL >Int 0
+```
+
+`#range(BM, START, WIDTH)` reads off `WIDTH` elements from `BM` beginning at position `START`, and converts it into an unsigned integer.
+The function interprets the range of bytes as little-endian.
+
+```k
+    syntax Int ::= #range ( Map , Int , Int ) [function]
+ // ----------------------------------------------------
+    rule #range(BM:Map, START, 0    ) => 0
+    rule #range(BM:Map, START, WIDTH) => #lookup(BM, START) +Int (#range(BM, START +Int 1, WIDTH -Int 1) *Int 256)
+      requires WIDTH >Int 0 [concrete]
```

Comment:
The Java prover will have a better time if the first rule is phrased as:

```
rule #range(BM, START, WIDTH) => 0 requires WIDTH ==Int 0
```

The reason is that a match on the LHS of a rule will be translated into `==K`, but here we can use the more specific `==Int` (which has it's own reasoning builtin).

Another option would be to mark *both* rules as `concrete`.

Reply:
Great explanation, thanks. It still makes sense to mark the second rule concrete anyway, correct?